components/style/values/specified/length.rs

```diff
@@ -458,6 +458,17 @@ impl Parse for Length {
     }
 }
 
+impl Either<Length, None_> {
+    #[inline]
+    #[allow(missing_docs)]
+    pub fn parse_non_negative_length(context: &ParserContext, input: &mut Parser) -> Result<Self, ()> {
+        if input.try(|input| None_::parse(context, input)).is_ok() {
+            return Ok(Either::Second(None_));
+        }
+        Length::parse_internal(input, AllowedNumericType::NonNegative).map(Either::First)
```

Comment:
Yes, that's because `T` is any type that implements the `Parse` trait, and so the compiler doesn't have any information about the value. What you should do is,

``` rust
input.try(|i| T::parse(context, i)).map(Either::Second).or_else(|_| {
    Length::parse_non_negative(input).map(Either::First)
})
```

... by which we're piping the parsed value into `Either`